core/include/vecmem/utils/impl/copy.ipp

```diff
@@ -395,6 +398,138 @@ copy::event_type copy::set_sizes(
+    }
+
+    // Initialize the "size variable(s)" correctly on the buffer.
+    if (data.size().ptr() != nullptr) {
+        memset(data.size(), 0);
+    }
+    VECMEM_DEBUG_MSG(3,
+                     "Prepared an SoA container of capacity %u "
+                     "for use on a device (layout: {%u, %p}, size: {%u, %p})",
+                     data.capacity(), data.layout().size(),
+                     static_cast<void*>(data.layout().ptr()),
+                     data.size().size(), static_cast<void*>(data.size().ptr()));
+
+    // Return a new event.
+    return create_event();
+}
+
+template <typename... VARTYPES>
+copy::event_type copy::memset(edm::view<edm::schema<VARTYPES...>> data,
+                              int value) const {
+
+    // For buffers, we can do this in one go.
+    if (data.payload().ptr() != nullptr) {
+        memset(data.payload(), value);
+    } else {
+        // Do the operation using the helper function, recursively.
+        memset_impl(data, value, std::index_sequence_for<VARTYPES...>{});
+    }
+
+    // Return a new event.
+    return create_event();
+}
+
+template <typename... VARTYPES1, typename... VARTYPES2>
+copy::event_type copy::operator()(
+    const edm::view<edm::schema<VARTYPES1...>>& from_view,
+    edm::view<edm::schema<VARTYPES2...>> to_view,
+    type::copy_type cptype) const {
+
+    // The input and output types are allowed to be different, but only by
+    // const-ness.
+    static_assert(sizeof...(VARTYPES1) == sizeof...(VARTYPES2),
+                  "Can only use compatible types in the copy");
+    static_assert(
+        std::conjunction_v<std::is_same<VARTYPES1, VARTYPES2>...> ||
+            std::conjunction_v<
+                edm::type::details::is_same_nc<VARTYPES1, VARTYPES2>...>,
+        "Can only use compatible types in the copy");
+
+    // First, handle the simple case, when both views have a contiguous memory
+    // layout.
+    if ((from_view.payload().ptr() != nullptr) &&
+        (to_view.payload().ptr() != nullptr) &&
+        (from_view.payload().size() == to_view.payload().size())) {
+
+        // If the "common size" is zero, we're done.
+        if (from_view.payload().size() == 0) {
+            return vecmem::copy::create_event();
+        }
+
+        // Let the user know what's happening.
+        VECMEM_DEBUG_MSG(2, "Performing simple SoA copy of %u bytes",
+                         from_view.payload().size());
+
+        // Copy the payload with a single copy operation.
+        operator()(from_view.payload(), to_view.payload(), cptype);
+
+        // If the target view is resizable, set its size.
+        if (to_view.size().ptr() != nullptr) {
+            // If the source is also resizable, the situation should be simple.
+            if (from_view.size().ptr() != nullptr) {
+                // Check that the sizes are the same.
+                if (from_view.size().size() != to_view.size().size()) {
```

Comment:
The naming is probably not too intuitive here. :frowning:

`vecmem::edm::view::size()` returns a `vecmem::data::vector_view<char>` object. It's a type-less block of memory, which describes "the size" of the container in some way. Again, as described in #254 a bit, the handling of resizable containers that have some jagged vectors and those that have none, go quite differently. For the types without jagged vectors, we just have a single `unsigned int` in memory that describes the size of the entire container. But when jagged vectors are involved, it's not the container that's "resizable", but all the individual jagged vectors by themselves. So we instead have a bunch of `unsigned int`-s allocated, one for each "inner vector" of every jagged vector variable.

`vecmem::edm::view::size()` returns a type-less view into the memory block that describes the "dynamic/resizable size" of the container.

If we reached this point in the code, where we checked already that the schema of the two views is "compatible" (so they should have the exact same variables except for const-ness), and both of them are resizable, then the two of them not using the same number of `unsigned int`-s to describe their size is some serious logic bug in the code somewhere. Not on the user's side, but on ours. Since the code should really never ever reach this point without some coding error in this project.

Maybe the exception message should reflect this a bit more. :thinking: But for now. throwing an exception seemed like the best thing to do here. :thinking: